edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
                 return true;
             }
 
-            // TODO: the following solution is to unblock the case when in a nested scenario a child edgehub
-            // subscribes in the name of a device/module. However changes needed to support unsubscribe
+            var deviceId = match.Groups["id1"].Success ? Option.Some<string>(match.Groups["id1"].Value) : Option.None<string>();
+            var moduleId = match.Groups["id2"].Success ? Option.Some<string>(match.Groups["id2"].Value) : Option.None<string>();
+
+            // For indirect clients the subscription change will be reported through edgeHub. That is a
+            // slower method to reconcile so we handle the direct case separately with a faster method
+            await moduleId.Filter(i => string.Equals(i, Constants.EdgeHubModuleId))
+                     .Match(
+                        _ => this.HandleNestedSubscriptionChanges(subscriptionList),
+                        () => this.HandleDirectSubscriptionChanges(deviceId, moduleId, subscriptionList));
+
+            return true;
+        }
+
+        async Task HandleDirectSubscriptionChanges(Option<string> deviceId, Option<string> moduleId, List<string> subscriptionList)
+        {
+            var identity = moduleId.Match(
+                               mod => deviceId.Match(
+                                         dev => this.identityProvider.Create(dev, mod),
+                                         () => throw new ArgumentException("Invalid topic structure for subscriptions - Module name matched but no Device name")),
+                               () => deviceId.Match(
+                                         dev => this.identityProvider.Create(dev),
+                                         () => throw new ArgumentException("Invalid topic structure for subscriptions - no Device name matched")));
+
+            var listener = default(IDeviceListener);
+            var maybeListener = await this.connectionRegistry.GetDeviceListenerAsync(identity, true);
```

Comment:
nit: rename GetOrCreateDeviceListeberAsync

Reply:
That sounds more correct, changed.